internal/infrastructure/datastore/hostconf_jwk_refresh.go

```diff
@@ -0,0 +1,234 @@
+			slog.Time("expires", hcjwk.ExpiresAt),
+			slog.Time("renewalAfter", renewAfter),
+		)
+		switch privstate {
+		case hostconf_jwk.ValidKey:
+			valid += 1
+			if hcjwk.ExpiresAt.Unix() >= renewAfter.Unix() {
+				log.Info("Valid Hostconf JWK")
+				create = false
+			} else {
+				log.Warn("Valid Hostconf JWK is after renewal threshold")
+			}
+		case hostconf_jwk.RevokedKey:
+			log.Info("Hostconf JWK is revoked")
+			revoked += 1
+		case hostconf_jwk.ExpiredKey:
+			log.Info("Hostconf JWK is expired")
+			expired += 1
+		default:
+			log.Info("Hostconf JWK is invalid")
+		}
+	}
+
+	slog.Info(
+		"Current JWKs in database",
+		slog.Int("total", len(hcjwks)),
+		slog.Int("valid", valid),
+		slog.Int("expired", expired),
+		slog.Int("revoked", revoked),
+	)
+
+	if create {
+		var newjwk *model.HostconfJwk
+
+		if valid == 0 {
+			slog.Warn("No valid JWK found in database")
+		} else {
+			slog.Warn("All valid JWKs expire in the renewal threshold period")
+		}
+
+		if newjwk, err = model.NewHostconfJwk(r.cfg.Secrets, expiresAt); err != nil {
+			return err
+		}
+		r.repository.InsertJWK(db, newjwk)
+		slog.Info(
+			"Created new hostconf JWK",
+			slog.String("kid", newjwk.KeyId),
+			slog.Time("expires", newjwk.ExpiresAt),
+		)
+	}
+
+	if tx.Commit(); tx.Error != nil {
+		return tx.Error
+	}
+	return nil
+}
+
+// Mark a JWK as revoked
+func (r *HostconfJwkDb) Revoke(kid string) (err error) {
+	var (
+		db    *gorm.DB
+		tx    *gorm.DB
+		hcjwk *model.HostconfJwk
+	)
+	db = NewDB(r.cfg)
+	defer Close(db)
+
+	if tx = db.Begin(); tx.Error != nil {
+		return tx.Error
+	}
+	defer tx.Rollback()
+
+	if hcjwk, err = r.repository.RevokeJWK(tx, kid); err != nil {
+		return err
+	}
+	slog.Info("Revoked JWK", slog.String("kid", hcjwk.KeyId))
+
+	if tx.Commit(); tx.Error != nil {
+		return tx.Error
+	}
+	return nil
+}
+
+// Purge and remove expired JWKs from database
+func (r *HostconfJwkDb) Purge() (err error) {
+	var (
+		db     *gorm.DB
+		tx     *gorm.DB
+		hcjwks []model.HostconfJwk
+	)
+	db = NewDB(r.cfg)
+	defer Close(db)
+
+	if tx = db.Begin(); tx.Error != nil {
+		return tx.Error
+	}
+	defer tx.Rollback()
+
+	if hcjwks, err = r.repository.PurgeExpiredJWKs(tx); err != nil {
+		return err
+	}
+	if len(hcjwks) > 0 {
+		slog.Info("Purged keys from DB", slog.Int("purged", len(hcjwks)))
+		for _, hcjwk := range hcjwks {
+			slog.Info(
+				"Purged key",
+				slog.String("kid", hcjwk.KeyId),
+				slog.Time("expires", hcjwk.ExpiresAt),
+			)
+		}
+	} else {
+		slog.Info("Nothing to purge")
+	}
+
+	if tx.Commit(); tx.Error != nil {
+		return tx.Error
+	}
+	return nil
+}
+
+// List all JWKs in database
+func (r *HostconfJwkDb) ListKeys() (err error) {
+	var (
+		db     *gorm.DB
+		tx     *gorm.DB
+		hcjwks []model.HostconfJwk
+		hcjwk  model.HostconfJwk
+	)
+	db = NewDB(r.cfg)
+	defer Close(db)
+
+	if tx = db.Begin(); tx.Error != nil {
+		return tx.Error
+	}
+	defer tx.Rollback()
+
+	r.timestamps()
```

Comment:
We are invoking `timestamps()` here... just to log some configuration (renewal threshold, and validity duration), the current time, and the current time + those values?

Unsure why this is useful in the context of the list command.  I think it would be more useful to have a per-JWK indication that the key is expired, revoked, or due for renewal (similar to what occurs in `refresh` command).

Reply:
Good point. I'm now included `expiresAfter` and `renewlAfter` information in the key log lines:

```
msg="Hostconf JWK configuration" now=2024-02-21T09:37:47.000Z renewAfter=2024-03-22T09:37:47.000Z renewThreshold=720h0m0s expiresAfter=2024-05-21T09:37:47.000Z validity=2160h0m0s
msg="JWKs in database" length=2
msg="Hostconf JWK" kid=mbDs3TIE publickey=revoked privatekey=revoked expires=2024-05-16T09:02:51.000Z expiresAfter=2024-05-21T09:37:47.000Z renewalAfter=2024-03-22T09:37:47.000Z
msg="Hostconf JWK" kid=PeBB1c46 publickey=valid privatekey=valid expires=2024-05-19T08:43:51.000Z expiresAfter=2024-05-21T09:37:47.000Z renewAfter=2024-03-22T09:37:47.000Z
```